File: server/internal/ws/protocol.go
package ws

import "github.com/stealthpte/server/internal/db"

type MsgType string

const (
	MsgHello    MsgType = "hello"
	MsgHelloAck MsgType = "hello_ack"
	MsgPing     MsgType = "ping"
	MsgPong     MsgType = "pong"
	MsgStats    MsgType = "stats"
	MsgTunnelAdd MsgType = "tunnel_add"
	MsgTunnelDel MsgType = "tunnel_del"
)

type HelloMsg struct {
	Type     MsgType `json:"type"`
	ClientID string  `json:"client_id"`
	Token    string  `json:"token"`
	Version  string  `json:"version"`
}

type HelloAckMsg struct {
	Type    MsgType      `json:"type"`
	Tunnels []TunnelInfo `json:"tunnels"`
}

type TunnelInfo struct {
	ID         int64  `json:"id"`
	Proto      string `json:"proto"`
	ServerPort int    `json:"server_port"`
	Target     string `json:"target"`
}

type TunnelAddMsg struct {
	Type       MsgType `json:"type"`
	ID         int64   `json:"id"`
	Proto      string  `json:"proto"`
	ServerPort int     `json:"server_port"`
	Target     string  `json:"target"`
}

type TunnelDelMsg struct {
	Type MsgType `json:"type"`
	ID   int64   `json:"id"`
}

type PingMsg struct {
	Type MsgType `json:"type"`
}

type PongMsg struct {
	Type MsgType `json:"type"`
}

type StatEntry struct {
	ID      int64 `json:"id"`
	RxBytes int64 `json:"rx_bytes"`
	TxBytes int64 `json:"tx_bytes"`
	Conns   int   `json:"conns"`
}

type StatsMsg struct {
	Type    MsgType     `json:"type"`
	Tunnels []StatEntry `json:"tunnels"`
}

func tunnelToInfo(t *db.Tunnel) TunnelInfo {
	return TunnelInfo{
		ID:         t.ID,
		Proto:      t.Proto,
		ServerPort: t.ServerPort,
		Target:     t.TargetAddr,
	}
}

File: client/src/protocol.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ServerMsg {
    HelloAck { tunnels: Vec<TunnelInfo> },
    TunnelAdd(TunnelAdd),
    TunnelDel { id: u64 },
    Ping,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ClientMsg {
    Hello {
        client_id: String,
        token: String,
        version: String,
    },
    Pong,
    Stats {
        tunnels: Vec<StatEntry>,
    },
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct TunnelInfo {
    pub id: u64,
    pub proto: String,
    pub server_port: u16,
    pub target: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct TunnelAdd {
    pub id: u64,
    pub proto: String,
    pub server_port: u16,
    pub target: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct StatEntry {
    pub id: u64,
    pub rx_bytes: u64,
    pub tx_bytes: u64,
    pub conns: u32,
}
